File: smash-ranks-android/app/src/main/java/com/garpr/android/views/PollFrequencyPreferenceView.java
package com.garpr.android.views;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Build;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.AlertDialog;
import android.util.AttributeSet;
import android.view.View;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.models.PollFrequency;
import com.garpr.android.preferences.Preference;
import com.garpr.android.preferences.RankingsPollingPreferenceStore;

import javax.inject.Inject;

public class PollFrequencyPreferenceView extends SimplePreferenceView implements
        DialogInterface.OnClickListener, Preference.OnPreferenceChangeListener<PollFrequency>,
        View.OnClickListener {

    @Inject
    RankingsPollingPreferenceStore mRankingsPollingPreferenceStore;


    public PollFrequencyPreferenceView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
    }

    public PollFrequencyPreferenceView(final Context context, final AttributeSet attrs,
            final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public PollFrequencyPreferenceView(final Context context, final AttributeSet attrs,
            final int defStyleAttr, final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        if (isInEditMode()) {
            return;
        }

        mRankingsPollingPreferenceStore.getPollFrequency().addListener(this);
        refresh();
    }

    @Override
    public void onClick(final DialogInterface dialog, final int which) {
        dialog.dismiss();

        final PollFrequency current = mRankingsPollingPreferenceStore.getPollFrequency().get();
        final PollFrequency selected = PollFrequency.values()[which];

        if (current == selected) {
            return;
        }

        mRankingsPollingPreferenceStore.getPollFrequency().set(selected);
        refresh();
    }

    @Override
    public void onClick(final View v) {
        final CharSequence[] items = new CharSequence[PollFrequency.values().length];

        for (int i = 0; i < PollFrequency.values().length; ++i) {
            items[i] = getResources().getText(PollFrequency.values()[i].getTextResId());
        }

        final PollFrequency current = mRankingsPollingPreferenceStore.getPollFrequency().get();
        final int checkedItem = current == null ? -1 : current.ordinal();

        new AlertDialog.Builder(getContext())
                .setSingleChoiceItems(items, checkedItem, this)
                .setTitle(R.string.poll_frequency)
                .show();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mRankingsPollingPreferenceStore.getPollFrequency().removeListener(this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        if (!isInEditMode()) {
            App.get().getAppComponent().inject(this);
            mRankingsPollingPreferenceStore.getPollFrequency().addListener(this);
        }

        setOnClickListener(this);
        setTitleText(R.string.poll_frequency);

        if (isInEditMode()) {
            return;
        }

        refresh();
    }

    @Override
    public void onPreferenceChange(final Preference<PollFrequency> preference) {
        if (ViewCompat.isAttachedToWindow(this)) {
            refresh();
        }
    }

    @Override
    public void refresh() {
        super.refresh();

        final PollFrequency pollFrequency = mRankingsPollingPreferenceStore.getPollFrequency().get();
        setDescriptionText(pollFrequency == null ? R.string.not_yet_set : pollFrequency.getTextResId());
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/PollFrequencyPreferenceView.kt
package com.garpr.android.views

import android.content.Context
import android.content.DialogInterface
import android.util.AttributeSet
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.garpr.android.R
import com.garpr.android.extensions.appComponent
import com.garpr.android.models.PollFrequency
import com.garpr.android.preferences.Preference
import com.garpr.android.preferences.RankingsPollingPreferenceStore
import javax.inject.Inject

class PollFrequencyPreferenceView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : SimplePreferenceView(context, attrs), DialogInterface.OnClickListener,
        Preference.OnPreferenceChangeListener<PollFrequency>, View.OnClickListener {

    @Inject
    protected lateinit var rankingsPollingPreferenceStore: RankingsPollingPreferenceStore


    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        if (isInEditMode) {
            return
        }

        rankingsPollingPreferenceStore.pollFrequency.addListener(this)
        refresh()
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        dialog.dismiss()

        val current = rankingsPollingPreferenceStore.pollFrequency.get()
        val selected = PollFrequency.values()[which]

        if (current == selected) {
            return
        }

        rankingsPollingPreferenceStore.pollFrequency.set(selected)
        refresh()
    }

    override fun onClick(v: View) {
        val items = arrayOfNulls<CharSequence>(PollFrequency.values().size)

        for (i in 0 until PollFrequency.values().size) {
            items[i] = resources.getText(PollFrequency.values()[i].textResId)
        }

        val current = rankingsPollingPreferenceStore.pollFrequency.get()
        val checkedItem = current?.ordinal ?: -1

        AlertDialog.Builder(context)
                .setSingleChoiceItems(items, checkedItem, this)
                .setTitle(R.string.poll_frequency)
                .show()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        rankingsPollingPreferenceStore.pollFrequency.removeListener(this)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (!isInEditMode) {
            appComponent.inject(this)
            rankingsPollingPreferenceStore.pollFrequency.addListener(this)
        }

        setOnClickListener(this)
        titleText = resources.getText(R.string.poll_frequency)

        if (isInEditMode) {
            return
        }

        refresh()
    }

    override fun onPreferenceChange(preference: Preference<PollFrequency>) {
        if (isAlive) {
            refresh()
        }
    }

    override fun refresh() {
        super.refresh()

        val pollFrequency = rankingsPollingPreferenceStore.pollFrequency.get()
        descriptionText = resources.getText(pollFrequency?.textResId ?: R.string.not_yet_set)
    }

}
